pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,412 @@
+		if err != nil {
+			klog.Warning("error removing leaseInformer eventhandler")
+		}
+		err = c.leaseCandidateInformer.Informer().RemoveEventHandler(c.leaseCandidateRegistration)
+		if err != nil {
+			klog.Warning("error removing leaseCandidateInformer eventhandler")
+		}
+	}()
+
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseRegistration.HasSynced, c.leaseCandidateRegistration.HasSynced) {
+		return
+	}
+
+	// This controller is leader elected and may start after informers have already started. List on startup.
+	lcs, err := c.leaseCandidateInformer.Lister().List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	for _, lc := range lcs {
+		c.processCandidate(lc)
+	}
+
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, c.runElectionWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func NewController(leaseInformer coordinationv1informers.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	c := &Controller{
+		leaseInformer:          leaseInformer,
+		leaseCandidateInformer: leaseCandidateInformer,
+		leaseClient:            leaseClient,
+		leaseCandidateClient:   leaseCandidateClient,
+
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[types.NamespacedName](), workqueue.TypedRateLimitingQueueConfig[types.NamespacedName]{Name: controllerName}),
+	}
+	leaseSynced, err := leaseInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.processLease(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.processLease(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.processLease(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	leaseCandidateSynced, err := leaseCandidateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.processCandidate(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.processCandidate(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.processCandidate(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	c.leaseRegistration = leaseSynced
+	c.leaseCandidateRegistration = leaseCandidateSynced
+	return c, nil
+}
+
+func (c *Controller) runElectionWorker(ctx context.Context) {
+	for c.processNextElectionItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextElectionItem(ctx context.Context) bool {
+	key, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+
+	completed, err := c.reconcileElectionStep(ctx, key)
+	utilruntime.HandleError(err)
+	if completed {
+		defer c.queue.AddAfter(key, requeueInterval)
```

Comment:
As discussed in Slack, I don't see why we should requeue after 5s. Election is complete. We know when the lease expires. If there is no change on the lease or on candidates (in either case we get an event), we know when the lease is due for election.

Reply:
Updated to return a requeueInterval instead so we have better control.